ludwig/benchmarking/profiler.py

```diff
@@ -0,0 +1,274 @@
+            self.launched = False
+            ex_type, ex_value, tb = sys.exc_info()
+            logging.error("Encountered exception when launching tracker thread.")
+            logging.error("".join(traceback.format_tb(tb)))
+            raise RuntimeError
+
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
+        """Stop profiling, postprocess and export resource usage metrics."""
+        try:
+            self.queue.put(STOP_MESSAGE)
+            self.t.join()
+            self.info = self.queue.get()
+            # recording in microseconds to be in line with torch profiler time recording.
+            self.info["end_time"] = time.perf_counter_ns() / 1000
+            self.launched = False
+        except Exception:
+            ex_type, ex_value, tb = sys.exc_info()
+            logging.error("Encountered exception when joining tracker thread.")
+            logging.error("".join(traceback.format_tb(tb)))
+        finally:
+            if self.use_torch_profiler:
+                self._ctx_exit_stack.close()
+                self._export_torch_metrics()
+            self._export_system_metrics()
+
+    def _export_system_metrics(self):
+        """Export system resource usage metrics (no torch operators)."""
+        self.info["total_execution_time"] = self.info.pop("end_time") - self.info.pop("start_time")
+        self.info["end_disk_usage"] = shutil.disk_usage(os.path.expanduser("~")).used
+        self.info["disk_footprint"] = self.info.pop("end_disk_usage") - self.info.pop("start_disk_usage")
+
+        all_keys = list(self.info.keys())
+        for key in all_keys:
+            if "cuda_" in key and "_memory_used" in key:
+                cuda_max_memory_key = key.replace("_memory_used", "_max_memory_used")
+                self.info[cuda_max_memory_key] = max(self.info[key], default=0)
+                cuda_average_memory_key = key.replace("_memory_used", "_average_memory_used")
+                if self.info[key]:
+                    self.info[cuda_average_memory_key] = mean(self.info.pop(key))
+                else:
+                    self.info.pop(key)
+                    self.info[cuda_average_memory_key] = 0
+
+        self.info["max_cpu_utilization"] = max(self.info["cpu_utilization"], default=0)
+        self.info["max_cpu_memory_usage"] = max(self.info["cpu_memory_usage"], default=0)
+
+        if self.info["cpu_utilization"]:
+            self.info["average_cpu_utilization"] = mean(self.info.pop("cpu_utilization"))
+        else:
+            self.info.pop("cpu_utilization")
+            self.info["average_cpu_utilization"] = 0
+        if self.info["cpu_memory_usage"]:
+            self.info["average_cpu_memory_usage"] = mean(self.info.pop("cpu_memory_usage"))
+        else:
+            self.info.pop("cpu_memory_usage")
+            self.info["average_cpu_memory_usage"] = 0
+
+        temp_dir = os.path.join(self.output_dir, "system_resource_usage", self.info["code_block_tag"])
```

Comment:
Is this a `temp_dir` or the actual output subdirectory? If the latter, maybe use `output_subdir`